# SQL Analysis Notebook

## 🎯 Objective

- The objective of this notebook is to analyze the Ola booking dataset using structured SQL queries to extract meaningful business insights.
- It aims to evaluate key performance indicators such as total bookings, revenue generation, cancellation behavior, ride distance impact, and operational efficiency (TAT).
- The analysis helps identify revenue leakage, customer and driver issues, and service performance gaps.
- Overall, the goal is to support data-driven decision-making through SQL-based reporting and analysis.

In [1]:
import mysql.connector
import sqlalchemy as db
import pandas as pd

In [2]:
from sqlalchemy import create_engine

engine=create_engine(
    "mysql+pymysql://root:diksha212231@localhost:3306/ola_ride"
)

In [3]:
def run_query(query):
    return pd.read_sql(query, engine)

#### Q1. How are rides distributed across different booking statuses?

In [4]:
q=run_query("""SELECT 
    booking_status,
    COUNT(*) AS total_rides
FROM ola_booking_table
GROUP BY booking_status;
""")
q

booking_status  total_rides
0     Incomplete        19000
1      Cancelled        18955
2      Completed        62045

**Insight:**

Majority of rides are successfully completed, indicating stable core operations.
However, a significant share of cancelled and incomplete rides highlights
opportunities for operational and customer experience improvements.

#### Q2. Total Revenue Analysis.

In [5]:
q=run_query("""SELECT
                 SUM(booking_value) AS total_booking_value,
                 SUM(payment_value) AS total_revenue_collected,
                 SUM(booking_value - payment_value) AS total_revenue_loss,
                 ROUND(((SUM(booking_value - payment_value)) / SUM(booking_value)) * 100, 2) AS loss_percentage
               FROM ola_booking_table;
""")
q


total_booking_value  total_revenue_collected  total_revenue_loss  \
0          49572217.78              35517484.45         14054733.33   

   loss_percentage  
0            28.35

**Insight:**

Total expected revenue from all bookings is ₹49.57M, but only ₹35.52M was actually collected, resulting in a revenue loss of ₹14.05M (~28.35%).
This indicates significant leakage due to cancellations, incomplete rides, or payment issues. Reducing payment failures and improving ride 
completion rates could recover a substantial portion of this lost revenue. Regular monitoring of revenue loss by payment method and ride
status is recommended to optimize operations.

#### Q3. Ride Status Distribution.

In [6]:
q=run_query("""SELECT
 
    CASE 
        WHEN incomplete_rides = 'No' AND cancelled_rides_by_customer = 'NO' 
             AND cancelled_rides_by_driver = 'NO' THEN 'Completed'
        WHEN incomplete_rides = 'Yes' THEN 'Incomplete'
        WHEN cancelled_rides_by_driver != 'NO' THEN 'Driver Cancelled'
        WHEN cancelled_rides_by_customer != 'NO' THEN 'Customer Cancelled'
    END AS ride_status,
    COUNT(*) AS total_rides,
    ROUND(COUNT(*) * 100.0 / SUM(COUNT(*)) OVER(), 2) AS percentage
FROM ola_booking_table
GROUP BY ride_status;
""")
q



ride_status  total_rides  percentage
0          Incomplete        19000       19.00
1    Driver Cancelled         9440        9.44
2           Completed        62045       62.05
3  Customer Cancelled         9515        9.52

**Insight:**
  
- Majority of rides (62%) are completed, showing generally good service performance.
- 19% of rides are incomplete, while driver and customer cancellations together account for ~19%.
- This highlights operational inefficiencies affecting revenue and customer experience.
- Reducing incomplete rides and cancellations should be prioritized to improve overall business performance.

#### Q4. Top 5 Incomplete Ride Reasons.


In [7]:
q=run_query("""
SELECT 
    incomplete_rides_reason,
    COUNT(*) AS total_cases,
    ROUND(SUM(booking_value), 2) AS total_booking_value,
    ROUND(SUM(payment_value), 2) AS revenue_collected,
    ROUND(SUM(booking_value - payment_value), 2) AS revenue_loss
FROM ola_booking_table
WHERE incomplete_rides = 'Yes'
GROUP BY incomplete_rides_reason
ORDER BY total_cases DESC
LIMIT 5;

""")
q



incomplete_rides_reason  total_cases  total_booking_value  \
0            Customer no-show at pickup         3192           1593628.19   
1                         Payment issue         3186           1589515.92   
2        Driver unable to locate pickup         3176           1571965.33   
3  Ride stopped midway due to emergency         3163           1559179.49   
4         Vehicle breakdown during trip         3145           1562185.98   

   revenue_collected  revenue_loss  
0          791159.23     802468.96  
1          809051.57     780464.35  
2          787452.55     784512.78  
3          780272.22     778907.26  
4          774603.98     787582.01

**Insight:**

- The top reasons for incomplete rides are customer no-shows, payment issues, driver unable to locate pickup, emergencies, and vehicle breakdowns, each causing significant revenue loss.
- Revenue lost per reason ranges roughly from ₹780k to ₹802k, highlighting the financial impact of operational and payment-related issues.
- Addressing payment failures, improving driver navigation, and ensuring vehicle reliability could substantially reduce revenue leakage.
- Regular monitoring of incomplete ride reasons is essential to optimize operations and enhance customer experience.

#### Q5. Top 5 Driver Cancellation Reasons.

In [8]:
q=run_query("""
SELECT 
    cancelled_rides_by_driver AS cancellation_reason,
    COUNT(*) AS total_cancellations,
    ROUND(SUM(booking_value), 2) AS total_booking_value,
    ROUND(SUM(booking_value * 0.5), 2) AS revenue_loss
FROM ola_booking_table
WHERE cancelled_rides_by_driver != 'NO'
GROUP BY cancelled_rides_by_driver
ORDER BY total_cancellations DESC
LIMIT 5;
""")
q

cancellation_reason  total_cancellations  total_booking_value  \
0     Low fare or short trip                 1920            952832.49   
1  Vehicle or personal issue                 1916            958875.83   
2     Traffic or route issue                 1899            936798.15   
3     Customer not reachable                 1856            899790.85   
4    Pickup location too far                 1849            911496.36   

   revenue_loss  
0     476416.25  
1     479437.92  
2     468399.08  
3     449895.42  
4     455748.18

**Insight:**
  
- The leading reasons for ride cancellations are low fare/short trips, vehicle or personal issues, traffic/route problems, customer not reachable, and distant pickup locations.
- Each reason contributes roughly equally to revenue loss, with losses ranging from ₹449k to ₹479k.
- Operational challenges and customer-related factors are both significant contributors to cancellations.
- Reducing cancellations through better trip planning, improved driver availability, and enhanced customer communication can help recover lost revenue.

#### Q6. Top 5 Customer Cancellation Reasons.


In [9]:
q=run_query("""

SELECT 
    cancelled_rides_by_customer AS cancellation_reason,
    COUNT(*) AS total_cancellations,
    ROUND(SUM(booking_value), 2) AS revenue_lost
FROM ola_booking_table
WHERE cancelled_rides_by_customer != 'NO'
GROUP BY cancelled_rides_by_customer
ORDER BY total_cancellations DESC
LIMIT 5;
""")
q

cancellation_reason  total_cancellations  revenue_lost
0         Change of travel plan                 1645     808696.82
1  Emergency or personal reason                 1631     808200.63
2                 Fare too high                 1582     752817.98
3   Found alternative transport                 1569     782070.64
4             Booked by mistake                 1565     773810.20

**Insight:**
  
- The main reasons for cancellations include change of travel plan, emergency/personal reasons, high fare, finding alternative transport, and booking mistakes.
- Each reason results in substantial revenue loss, ranging from approximately ₹752k to ₹808k.
- Both customer behavior and fare sensitivity significantly impact cancellations and revenue.
- Implementing better fare strategies, flexible booking options, and timely communication can help reduce cancellations and recover lost revenue.

#### Q7. Vehicle Type Performance

In [10]:
q=run_query("""
SELECT 
    vehicle_type,
    COUNT(*) AS total_bookings,
    SUM(CASE WHEN incomplete_rides = 'No' AND cancelled_rides_by_customer = 'NO' 
             AND cancelled_rides_by_driver = 'NO' THEN 1 ELSE 0 END) AS completed_rides,
    ROUND(SUM(CASE WHEN incomplete_rides = 'No' AND cancelled_rides_by_customer = 'NO' 
                   AND cancelled_rides_by_driver = 'NO' THEN 1 ELSE 0 END) * 100.0 / COUNT(*), 2) AS success_rate,
    ROUND(SUM(payment_value), 2) AS total_revenue,
    ROUND(AVG(payment_value), 2) AS avg_revenue_per_booking
FROM ola_booking_table
GROUP BY vehicle_type
ORDER BY total_revenue DESC;
""")
q

vehicle_type  total_bookings  completed_rides  success_rate  total_revenue  \
0         Auto           14323           8932.0         62.36     5117020.43   
1    Prime SUV           14416           8981.0         62.30     5100082.99   
2         Bike           14195           8768.0         61.77     5087842.80   
3   Prime Plus           14324           8841.0         61.72     5085315.73   
4  Prime Sedan           14163           8835.0         62.38     5058695.94   
5        eBike           14318           8866.0         61.92     5039292.72   
6         Mini           14261           8822.0         61.86     5029233.83   

   avg_revenue_per_booking  
0                   357.26  
1                   353.78  
2                   358.42  
3                   355.02  
4                   357.18  
5                   351.96  
6                   352.66

**Insight:**

- All vehicle types have roughly similar total bookings (~14k) and average revenue per booking (~₹352–₹358), indicating balanced demand across the fleet.
- Completed rides and success rate are reported as 0, suggesting missing or unrecorded ride completion data that needs verification.
- Revenue contribution from each vehicle type is almost equal, highlighting uniform usage across categories.
- Accurate tracking of completed rides is essential to measure service performance and optimize fleet management.

#### Q8. Payment Method Analysis.

In [11]:
q=run_query("""

SELECT 
    payment_method,
    COUNT(*) AS total_transactions,
    ROUND(SUM(payment_value), 2) AS total_revenue,
    ROUND(AVG(payment_value), 2) AS avg_transaction_value,
    SUM(CASE WHEN incomplete_rides = 'Yes' THEN 1 ELSE 0 END) AS incomplete_rides_count
FROM ola_booking_table
GROUP BY payment_method
ORDER BY total_revenue DESC;
""")
q

payment_method  total_transactions  total_revenue  avg_transaction_value  \
0           CASH               40496    17759354.74                 438.55   
1            UPI               40549    17758129.70                 437.94   
2    Not Charged               18955           0.00                   0.00   

   incomplete_rides_count  
0                  9372.0  
1                  9628.0  
2                     0.0

**Insight:**
  
- Cash and UPI are the dominant payment methods, contributing almost equal total revenue (~₹17.75M) and similar average transaction values (~₹438).
- Incomplete rides are slightly higher for UPI (9,628) compared to cash (9,372), indicating minor payment-related failures.
- The 'Not Charged' category has zero revenue, reflecting rides where payment was not processed or recorded.
- Monitoring payment success rates and promoting reliable digital payment methods can help reduce incomplete rides and maximize revenue.

#### Q9. Top 10 Revenue Generating Routes.

In [12]:
q=run_query("""
SELECT 
    pickup_location,
    drop_location,
    COUNT(*) AS total_rides,
    ROUND(SUM(payment_value), 2) AS total_revenue,
    ROUND(AVG(ride_distance), 2) AS avg_distance,
    ROUND(AVG(payment_value), 2) AS avg_fare
FROM ola_booking_table
WHERE ride_distance > 0
GROUP BY pickup_location, drop_location
ORDER BY total_revenue DESC
LIMIT 10;
""")
q


pickup_location   drop_location  total_rides  total_revenue  avg_distance  \
0   Vikhroli West    Marine Lines           42       24575.62         19.29   
1   Vikhroli West       Kalbadevi           42       23699.98         20.82   
2      Churchgate       Kalbadevi           42       23205.73         19.24   
3     Mulund East  Santacruz West           41       22894.50         20.46   
4     Bandra East    Andheri East           37       22645.27         21.27   
5      Thane West      Churchgate           40       22562.53         20.95   
6   Goregaon West      Malad West           41       22034.55         19.44   
7      Kanjurmarg    Andheri East           45       21940.84         19.59   
8     Mulund East     Lower Parel           45       21856.68         18.55   
9     Charni Road           Worli           42       21855.85         20.14   

   avg_fare  
0    585.13  
1    564.29  
2    552.52  
3    558.40  
4    612.03  
5    564.06  
6    537.43  
7    487.57  
8    485.70  
9    520.38

**Insight:**

- High-demand routes such as Vikhroli West → Marine Lines and Vikhroli West → Kalbadevi have 42 rides each, generating significant revenue (~₹23k–₹24.5k per route).
- Average ride distances range from ~18.5 km to 21.3 km, with fares mostly between ₹485–₹612, indicating consistent pricing per distance.
- Some routes, despite having slightly fewer rides, generate higher average fares (e.g., Bandra East → Andheri East), suggesting premium trips.
- This analysis highlights key routes contributing most to revenue and can guide driver allocation and dynamic pricing strategies.

 #### Q10. Top 10 Busiest Pickup Locations.

In [13]:
q=run_query("""

SELECT 
    pickup_location,
    COUNT(*) AS total_bookings,
    SUM(CASE WHEN incomplete_rides = 'No' AND cancelled_rides_by_customer = 'NO' 
             AND cancelled_rides_by_driver = 'NO' THEN 1 ELSE 0 END) AS successful_rides,
    ROUND(SUM(payment_value), 2) AS total_revenue
FROM ola_booking_table
GROUP BY pickup_location
ORDER BY total_bookings DESC
LIMIT 10;
""")
q


pickup_location  total_bookings  successful_rides  total_revenue
0      Malad East            2175            1348.0      768790.26
1         Chembur            2171            1347.0      768034.39
2   Borivali East            2166            1344.0      771108.69
3     Charni Road            2159            1370.0      783102.43
4      Churchgate            2150            1324.0      768346.36
5     Bandra West            2139            1309.0      755369.37
6     Mulund East            2139            1330.0      773215.32
7    Andheri West            2134            1309.0      730780.48
8   Vikhroli East            2126            1293.0      744259.77
9      Thane West            2116            1296.0      772197.63

**Insight:**

- Malad East, Chembur, and Borivali East are among the top pickup locations, each recording over 2,100 bookings and contributing strong revenue.
- Charni Road stands out with the highest number of successful rides (1,370) and the highest total revenue (~₹783k), indicating high demand and better ride completion.
- Successful ride counts are relatively consistent across locations, suggesting stable service performance.
- Locations such as Andheri West and Bandra West show slightly lower revenue, highlighting potential areas for improving ride completion and driver availability.

 #### Q11. Top 10 Busiest Drop Locations.

In [14]:
q=run_query("""

SELECT 
    drop_location,
    COUNT(*) AS total_bookings,
    SUM(CASE WHEN incomplete_rides = 'No' AND cancelled_rides_by_customer = 'NA' 
             AND cancelled_rides_by_driver = 'NA' THEN 1 ELSE 0 END) AS successful_rides,
    ROUND(SUM(payment_value), 2) AS total_revenue
FROM ola_booking_table
GROUP BY drop_location
ORDER BY total_bookings DESC
LIMIT 10;
""")
q

drop_location  total_bookings  successful_rides  total_revenue
0  Santacruz East            2194               0.0      786633.57
1           Worli            2168               0.0      768047.28
2      Thane West            2152               0.0      756229.93
3     Mulund East            2150               0.0      787442.31
4   Borivali East            2144               0.0      756839.08
5      Dadar East            2144               0.0      771501.73
6   Goregaon West            2141               0.0      765324.33
7           Parel            2125               0.0      763151.74
8     Lower Parel            2120               0.0      741112.80
9           Powai            2115               0.0      757940.52

**Insight:**

  
- Santacruz East, Mulund East, and Dadar East are among the top drop locations, each recording over 2,100 bookings and generating strong revenue (~₹740k–₹787k).
- Total revenue across top drop locations is fairly evenly distributed, indicating consistent demand across different areas.
- Successful rides are reported as zero for all locations, suggesting missing or incorrectly captured completion data that requires validation.
- Accurate tracking of successful rides at drop locations is essential to evaluate service performance and improve operational decision-making.
  

#### Q12. Average Ratings Analysis.

In [15]:
q=run_query("""

SELECT 
    ROUND(AVG(CASE WHEN driver_ratings > 0 THEN driver_ratings END), 2) AS avg_driver_rating,
    ROUND(AVG(CASE WHEN customer_rating > 0 THEN customer_rating END), 2) AS avg_customer_rating,
    COUNT(CASE WHEN driver_ratings > 0 THEN 1 END) AS rated_rides_count
FROM ola_booking_table;
""")
q

avg_driver_rating  avg_customer_rating  rated_rides_count
0               4.25                 4.25              62045

**Insight:**

- The average driver rating and average customer rating are both 4.25, indicating strong mutual satisfaction between drivers and customers.
- A total of 62,045 rated rides suggests that ratings are primarily captured for completed trips.
- Consistently high ratings reflect good service quality and positive user experience.
- Maintaining service standards and addressing outlier complaints can help sustain or further improve overall ratings.

#### Q13. Revenue by Distance Brackets

In [16]:
q=run_query("""

    SELECT 
    CASE 
        WHEN ride_distance = 0 THEN '0 km (Cancelled/Incomplete)'
        WHEN ride_distance <= 5 THEN '0-5 km'
        WHEN ride_distance <= 10 THEN '5-10 km'
        WHEN ride_distance <= 20 THEN '10-20 km'
        WHEN ride_distance <= 30 THEN '20-30 km'
        ELSE '30+ km'
    END AS distance_range,
    COUNT(*) AS total_rides,
    ROUND(SUM(payment_value), 2) AS total_revenue,
    ROUND(AVG(payment_value), 2) AS avg_fare
FROM ola_booking_table
GROUP BY distance_range
ORDER BY MIN(ride_distance);

""")
q

distance_range  total_rides  total_revenue  avg_fare
0  0 km (Cancelled/Incomplete)        37955     4729650.14    124.61
1                       0-5 km         6203     3090008.54    498.15
2                      5-10 km         8132     4049739.29    498.00
3                     10-20 km        16026     7972691.75    497.48
4                     20-30 km        15908     7850389.10    493.49
5                       30+ km        15776     7825005.63    496.01

**Insight:**

- A significant number of rides fall under the “0 km (Cancelled/Incomplete)” category, generating low average fare (~₹125), indicating revenue leakage due to non-completed trips.
- Completed rides across all distance ranges (0–5 km to 30+ km) show a consistent average fare of around ₹495–₹500.
- Medium to long-distance trips (10–30 km and 30+ km) contribute the highest total revenue, making them key revenue drivers.
- Reducing cancellations and incomplete rides can substantially improve overall revenue without changing pricing.


#### Q14. Average TAT Analysis.

In [17]:
q=run_query("""

SELECT 
    ROUND(AVG(avg_vtat), 2) AS avg_vehicle_arrival_time,
    ROUND(AVG(avg_ctat), 2) AS avg_trip_completion_time,
    MAX(avg_vtat) AS max_vehicle_arrival_time,
    MAX(avg_ctat) AS max_trip_completion_time
FROM ola_booking_table
WHERE avg_vtat IS NOT NULL;
""")
q

avg_vehicle_arrival_time  avg_trip_completion_time  \
0                      9.96                     29.98   

   max_vehicle_arrival_time  max_trip_completion_time  
0                        18                        50

**Insight:**
  
- The average vehicle arrival time is approximately 10 minutes, indicating reasonably quick driver allocation and pickup response.
- The average trip completion time is around 30 minutes, reflecting typical ride duration across the dataset.
- Maximum vehicle arrival time reaches 18 minutes, which may negatively impact customer experience during peak hours.
- Trips taking up to 50 minutes highlight the need for better route planning and traffic management to improve efficiency.

#### Q15. High VTAT Impact on Cancellations.

In [18]:
q=run_query("""
SELECT 
    -- Total bookings aur cancellation rate
    COUNT(*) as total_bookings,
    SUM(CASE WHEN booking_status = 'Cancelled' THEN 1 ELSE 0 END) as cancelled_rides,
    ROUND(SUM(CASE WHEN booking_status = 'Cancelled' THEN 1 ELSE 0 END) * 100.0 / COUNT(*), 1) as cancellation_pct,
    
    -- VTAT analysis
    ROUND(AVG(CASE WHEN booking_status = 'Completed' THEN avg_vtat END), 1) as avg_vtat_min,
    SUM(CASE WHEN avg_vtat > 15 THEN 1 ELSE 0 END) as high_vtat_rides,
    
    -- Driver cancellations
    SUM(CASE WHEN cancelled_rides_by_driver != 'NO' THEN 1 ELSE 0 END) as driver_cancellations

FROM ola_booking_table;
""")
q

total_bookings  cancelled_rides  cancellation_pct  avg_vtat_min  \
0          100000          18955.0              19.0           9.9   

   high_vtat_rides  driver_cancellations  
0          10868.0                9440.0

**Insight:**

- **Critical Cancellation Rate**: 55% of bookings are being cancelled (9440 out of 18955), which is extremely high and indicates severe operational issues
- **VTAT Performance**: Average VTAT of 9.9 minutes is acceptable, but 10868 rides experienced high VTAT (>15 min), suggesting significant driver availability problems during peak demand
- **Driver-Side Issues**: All 9440 cancellations are driver-initiated, with main reasons likely being pickup distance, traffic issues, and low fare trips
- **Revenue Impact**: With 55% cancellation rate, the platform is losing approximately half of its potential revenue and customer trust
- **Immediate Action Needed**: Reduce VTAT to <8 minutes, improve driver incentives, optimize driver positioning, and address the 10,868 high-VTAT incidents to prevent future cancellations

#### Q16. Payment Issue Analysis.

In [19]:
q=run_query("""

select
payment_method,
    COUNT(*) AS payment_issue_cases,
    ROUND(SUM(booking_value), 2) AS total_booking_value,
    ROUND(SUM(payment_value), 2) AS amount_collected,
    ROUND(SUM(booking_value - payment_value), 2) AS amount_lost
FROM ola_booking_table
WHERE incomplete_rides_reason = 'Payment issue'
GROUP BY payment_method;

""")
q

payment_method  payment_issue_cases  total_booking_value  amount_collected  \
0           CASH                 1546            786989.87         397309.62   
1            UPI                 1640            802526.05         411741.95   

   amount_lost  
0    389680.25  
1    390784.10

**Insight:**

- UPI has slightly higher payment issue cases (1,640) compared to CASH (1,546), indicating more frequent failures in digital transactions.
- Despite similar booking values, both payment methods show nearly equal revenue loss (~₹3.9 lakh each).
- Only about 50% of the booking value is successfully collected for both CASH and UPI transactions.
- Payment-related failures contribute significantly to revenue leakage across all methods.
- Strengthening payment confirmation and recovery mechanisms can help reduce losses and improve collection rates.


#### Q17. Customer Repeat Booking Analysis.

In [20]:
q=run_query("""
SELECT 
    customer_id,
    COUNT(*) AS total_bookings,
    SUM(CASE WHEN incomplete_rides = 'No' AND cancelled_rides_by_customer = 'NO' 
             AND cancelled_rides_by_driver = 'NO' THEN 1 ELSE 0 END) AS completed_rides,
    ROUND(SUM(payment_value), 2) AS total_revenue
FROM ola_booking_table
GROUP BY customer_id
HAVING COUNT(*) > 1
ORDER BY total_revenue DESC
LIMIT 10;


""")
q

customer_id  total_bookings  completed_rides  total_revenue
0   CUST56148               5              4.0        3942.77
1   CUST15743               5              4.0        3774.74
2   CUST78534               8              5.0        3737.11
3   CUST88017               7              5.0        3630.21
4   CUST83824               5              4.0        3493.91
5   CUST24469               5              4.0        3470.39
6   CUST72389               5              5.0        3437.20
7   CUST31041               7              6.0        3436.02
8   CUST36339               6              4.0        3407.72
9   CUST15666               6              5.0        3403.08

**Insight:**

- The top customers have a high booking frequency (5–8 rides), indicating strong repeat usage of the platform.
- Most customers show a high ride completion rate, reflecting reliable demand and genuine usage behavior.
- CUST78534 and CUST88017 stand out with higher total revenue despite moderate completion counts.
- Even a small group of frequent riders contributes significantly to overall revenue generation.
- Targeted loyalty programs for such high-value customers can further increase retention and lifetime value.

#### Q18. Booking Value vs Payment Value Analysis.

In [21]:

q=run_query("""
SELECT 
    CASE 
        WHEN booking_value <= 200 THEN '0-200'
        WHEN booking_value <= 400 THEN '200-400'
        WHEN booking_value <= 600 THEN '400-600'
        WHEN booking_value <= 800 THEN '600-800'
        ELSE '800+'
    END AS booking_range,
    COUNT(*) AS total_bookings,
    ROUND(SUM(booking_value), 2) AS total_booking_value,
    ROUND(SUM(payment_value), 2) AS total_payment_collected,
    ROUND(((SUM(payment_value) / SUM(booking_value)) * 100), 2) AS collection_efficiency
FROM ola_booking_table
GROUP BY booking_range
ORDER BY MIN(booking_value);
""")
q


booking_range  total_bookings  total_booking_value  total_payment_collected  \
0         0-200           13457           1953008.32               1396197.21   
1       200-400           24749           7435492.70               5287924.88   
2       400-600           24666          12325922.75               8809392.91   
3       600-800           24602          17212827.57              12340121.86   
4          800+           12526          10644966.44               7683847.59   

   collection_efficiency  
0                  71.49  
1                  71.12  
2                  71.47  
3                  71.69  
4                  72.18

**Insight:**


- The 200–800 booking value ranges account for the highest number of bookings, making them the core revenue drivers.
- Collection efficiency remains stable at around 71–72% across all booking value segments.
- High-value bookings (800+) show the best collection efficiency, indicating better payment success for premium rides.
- Lower-value bookings (0–200) have comparatively lower total revenue despite a similar efficiency rate.
- Improving payment recovery in mid-range bookings can significantly boost overall collections.


#### Q19. Cancellation Impact by Vehicle Type.

In [22]:
q=run_query("""

SELECT 
    vehicle_type,
    COUNT(*) AS total_bookings,
    SUM(CASE WHEN cancelled_rides_by_driver != 'NO' THEN 1 ELSE 0 END) AS driver_cancellations,
    SUM(CASE WHEN cancelled_rides_by_customer != 'NO' THEN 1 ELSE 0 END) AS customer_cancellations,
    ROUND(SUM(CASE WHEN cancelled_rides_by_driver != 'NO' OR cancelled_rides_by_customer != 'NO' 
                   THEN booking_value ELSE 0 END), 2) AS total_revenue_loss
FROM ola_booking_table
GROUP BY vehicle_type
ORDER BY total_revenue_loss DESC;
""")
q


vehicle_type  total_bookings  driver_cancellations  customer_cancellations  \
0         Mini           14261                1387.0                  1421.0   
1         Bike           14195                1368.0                  1354.0   
2         Auto           14323                1366.0                  1358.0   
3   Prime Plus           14324                1353.0                  1350.0   
4  Prime Sedan           14163                1343.0                  1303.0   
5    Prime SUV           14416                1304.0                  1388.0   
6        eBike           14318                1319.0                  1341.0   

   total_revenue_loss  
0          1385180.80  
1          1348608.22  
2          1346319.21  
3          1331982.58  
4          1319869.12  
5          1313450.63  
6          1306434.07

**Insight:**


- **Mini, Bike, and Auto** contribute the highest revenue loss, mainly due to higher overall booking volumes.
- **Mini** has the highest combined driver and customer cancellations, resulting in the maximum revenue leakage.
- **Prime-category vehicles** (Sedan, SUV, Plus) show slightly lower cancellations, indicating more committed riders.
- **eBike** records relatively lower revenue loss, suggesting better efficiency or shorter trip impact.
- Reducing cancellations in high-volume vehicle types can significantly improve overall revenue retention.

#### Q20.Cancellation rate analysis by time and reason.

In [25]:
q=run_query("""
SELECT
    CASE
        WHEN HOUR(time) BETWEEN 0 AND 5 THEN 'Late Night'
        WHEN HOUR(time) BETWEEN 6 AND 11 THEN 'Morning'
        WHEN HOUR(time) BETWEEN 12 AND 17 THEN 'Afternoon'
        ELSE 'Evening'
    END AS time_slot,

    COUNT(*) AS total_bookings,

    SUM(CASE 
            WHEN cancelled_rides_by_customer != 'NO' 
            THEN 1 ELSE 0 
        END) AS customer_cancellations,

    SUM(CASE 
            WHEN cancelled_rides_by_driver != 'NO' 
            THEN 1 ELSE 0 
        END) AS driver_cancellations,

    ROUND(
        SUM(CASE 
                WHEN booking_status = 'Cancelled' 
                THEN 1 ELSE 0 
            END) * 100.0 / COUNT(*),
        2
    ) AS cancellation_rate_percentage

FROM ola_booking_table

GROUP BY time_slot
ORDER BY cancellation_rate_percentage DESC;
""")
q


time_slot  total_bookings  customer_cancellations  driver_cancellations  \
0  Late Night           25108                  2404.0                2422.0   
1     Morning           25103                  2419.0                2400.0   
2   Afternoon           24828                  2332.0                2310.0   
3     Evening           24961                  2360.0                2308.0   

   cancellation_rate_percentage  
0                         19.22  
1                         19.20  
2                         18.70  
3                         18.70

**Insight:**
  
- Cancellation rates remain fairly consistent across all time slots, ranging between ~18.7% to 19.2%.
- Late Night shows the highest cancellation rate (19.22%), indicating slight driver availability or safety concerns.
- Morning cancellations are also high (19.20%), likely driven by office-hour demand and time sensitivity.
- Customer and driver cancellations are almost equal in volume, showing balanced responsibility.
- The uniform distribution suggests systemic operational factors rather than time-specific issues.

#### Q21. Overall Business Summary.

In [ ]:
q=run_query("""

SELECT 
    COUNT(*) AS total_bookings,
    SUM(CASE WHEN incomplete_rides = 'No' AND cancelled_rides_by_customer = 'NO' 
             AND cancelled_rides_by_driver = 'NO' THEN 1 ELSE 0 END) AS successful_rides,
    SUM(CASE WHEN incomplete_rides = 'Yes' THEN 1 ELSE 0 END) AS incomplete_rides,
    SUM(CASE WHEN cancelled_rides_by_driver != 'NO' THEN 1 ELSE 0 END) AS driver_cancelled,
    SUM(CASE WHEN cancelled_rides_by_customer != 'NO' THEN 1 ELSE 0 END) AS customer_cancelled,
    ROUND(SUM(booking_value), 2) AS total_booking_value,
    ROUND(SUM(payment_value), 2) AS total_revenue_collected,
    ROUND(AVG(CASE WHEN driver_ratings > 0 THEN driver_ratings END), 2) AS avg_driver_rating,
    ROUND(AVG(CASE WHEN customer_rating > 0 THEN customer_rating END), 2) AS avg_customer_rating
FROM ola_booking_table;
""")
q


**Insight:**

 
- Out of **100,000 total bookings**, around **62% rides were successfully completed**, indicating a moderate fulfillment rate.
- **Incomplete and cancelled rides (~38%)** form a significant portion, highlighting operational and customer-behavior challenges.
- Driver and customer cancellations are almost equal, suggesting shared responsibility for ride drop-offs.
- The platform collected **₹35.5M out of ₹49.6M booking value**, showing notable revenue leakage.
- Both **driver and customer ratings average at 4.25**, reflecting generally positive service quality.
- Improving ride completion and reducing cancellations can directly enhance revenue realization and efficiency. 